In [1]:
import warnings                                  # `do not disturbe` mode
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.graph_objects as go

In [2]:
df = pd.read_csv('FAOSTAT_downloaded20210119.csv')
foods = pd.read_csv('foods.csv')

In [3]:
#Bind old (-2013) and new time series (2014-): Food Balances and Producer prices
df.Domain.replace('Food Balances (old methodology and population)', 'FoodBalances', inplace=True)
df.Domain.replace('New Food Balances', 'FoodBalances', inplace=True)
df.Domain.replace('Producer Prices', 'ProducerPrices', inplace=True)
df.Domain.replace('Producer Prices (old series)', 'ProducerPrices', inplace=True)
df.Domain.replace('Food Supply - Crops Primary Equivalent', 'FoodSupply', inplace=True)
df.Domain.replace('Crops', 'CropProduction', inplace=True)
df.Domain.replace('Value of Agricultural Production', 'ValueAgProduction', inplace=True)

#merge with dictionary including "Food groups". Category 2 based on HDDS.
#if you want to filter out non-food, processed foods and beverages, select 'Crop' and 'CropProducts'
df = df.merge(foods, how='left', on='Item')

In [ ]:
#df.columns
df.Element.unique()
#df[df.Domain == 'ValueAgProduction'].Element.unique()


In [28]:
df1 = df[(df.Domain == 'FoodBalances') & 
         ((df.Category1 == 'Crop')|(df.Category1 == 'CropProducts')) &
         (df.Year >= 1991) & (df.Year <= 2019)]
df1 = df1[['Domain', 'Area', 'Year', 'Element', 'Value', 'name']]
df1 = df1.rename(columns={'name' : 'Item'})

a = df1[(df1.Element == 'Production') ].groupby(['Item', 'Area']).mean().reset_index()
a = a.rename(columns={'Value':'Production'})
b = df1[ (df1.Element == 'Food')].groupby(['Item', 'Area']).mean().reset_index()
b = b.rename(columns={'Value':'Food'})
d = df1[(df1.Element == 'Import Quantity')].groupby(['Item', 'Area']).mean().reset_index()
d = d.rename(columns={'Value':'Import'})
e = df1[(df1.Element == 'Export Quantity')].groupby(['Item', 'Area']).mean().reset_index()
e = e.rename(columns={'Value':'Export'})
f = df1[(df1.Element == 'Domestic supply quantity')].groupby(['Item', 'Area']).mean().reset_index()
f = f.rename(columns={'Value':'DomesticSupply'})
g = df1[(df1.Element == 'Food supply quantity (kg/capita/yr)')].groupby(['Item', 'Area']).mean().reset_index()
g = g.rename(columns={'Value':'FoodPerCapita'})

df3 = df[(df.Domain == 'ProducerPrices') & 
         ((df.Category1 == 'Crop')|(df.Category1 == 'CropProducts')) &
         (df.Year >= 1991) & (df.Year <= 2019)] 
df3 = df3[['Domain', 'Area', 'Year', 'Element', 'Value', 'name']]
df3 = df3.rename(columns={'name' : 'Item'})
h = df3[(df3.Element == 'Producer Price Index (2014-2016 = 100)') ].groupby(['Item', 'Area']).mean().reset_index()
h = h.rename(columns = {'Value':'PriceIndex'})

df4 = df[(df.Domain == 'ValueAgProduction') & 
         ((df.Category1 == 'Crop')|(df.Category1 == 'CropProducts')) &
         (df.Year >= 1991) & (df.Year <= 2019)] 
df4 = df4[['Domain', 'Area', 'Year', 'Element', 'Value', 'name']]
df4 = df4.rename(columns={'name' : 'Item'})
i = df4[(df4.Element == 'Gross Production Value (constant 2014-2016 million US$)') ].groupby(['Item', 'Area']).mean().reset_index()
i = i.rename(columns = {'Value':'GrossProductionValue'})

df5 = df[(df.Domain == 'CropProduction') & 
         ((df.Category1 == 'Crop')|(df.Category1 == 'CropProducts')) &
         (df.Year >= 1991) & (df.Year <= 2019)] 
df5 = df5[['Domain', 'Area', 'Year', 'Element', 'Value', 'name']]
df5 = df5.rename(columns={'name' : 'Item'})
j = df5[(df5.Element == 'Area harvested') ].groupby(['Item', 'Area']).mean().reset_index()
j = j.rename(columns = {'Value':'Area harvested'})

a.drop('Year', axis=1, inplace=True)
b.drop('Year', axis=1, inplace=True)
g.drop('Year', axis=1, inplace=True)
h.drop('Year', axis=1, inplace=True)
i.drop('Year', axis=1, inplace=True)
j.drop('Year', axis=1, inplace=True)

df2 = a.merge(b, how='outer', on=['Item', 'Area'])
df2 = df2.merge(g, how= 'outer', on=['Item', 'Area'])
df2 = df2.merge(h, how= 'outer', on=['Item', 'Area'])
df2 = df2.merge(i, how= 'outer', on=['Item', 'Area'])
df2 = df2.merge(j, how= 'outer', on=['Item', 'Area'])

df2['FoodSelfSufficiency'] = df2.Production / df2.Food  *100
df2['FoodSelfSufficiency'] = df2['FoodSelfSufficiency'].round(1)
df2['FoodPerCapita'] = df2['FoodPerCapita'].round(1)
df2['PriceIndex'] = df2['PriceIndex'].round(1)
df2['GrossProductionValue'] = df2['GrossProductionValue'].round(1)
df2['Production'] = df2['Production'].round(1)
df2['Food'] = df2['Food'].round(1)

df2.sort_values('Area harvested', ascending=False).head(20)


Item Area  Production    Food  FoodPerCapita  \
57                    Maize  TZA      3839.6  2544.4           65.5   
55                    Maize  KEN      2879.1  2975.2           83.1   
287    Plantains and others  UGA         NaN     NaN            NaN   
17                  Cassava  TZA      5597.4  4828.2          132.7   
58                    Maize  UGA      1618.4   996.0           32.6   
110             Rice, paddy  TZA       964.1   873.9           20.8   
121                 Sorghum  TZA       686.7   289.8            7.6   
18                  Cassava  UGA      3549.0  2410.0           85.1   
132          Sunflower seed  TZA       513.9    35.0            0.7   
200  Groundnuts, with shell  TZA         NaN     NaN            NaN   
137          Sweet potatoes  UGA      2124.7  1909.5           69.4   
27                 Coconuts  TZA       461.0   245.0            6.0   
136          Sweet potatoes  TZA      1690.6  1287.7           29.3   
292             Seed cotton  TZA         NaN     NaN            NaN   
4                   Bananas  RWA      2318.1  2188.6          254.8   
117             Sesame seed  TZA       172.1     NaN            NaN   
122                 Sorghum  UGA       384.3   124.1            4.5   
5                   Bananas  TZA      1887.4  1298.4           29.8   
267           Coffee, green  UGA         NaN     NaN            NaN   
62                   Millet  UGA       465.1   337.3           13.3   

     PriceIndex  GrossProductionValue  Area harvested  FoodSelfSufficiency  
57         58.7                 723.9    2.654626e+06                150.9  
55         60.3                 927.5    1.764822e+06                 96.8  
287         NaN                   NaN    1.299446e+06                  NaN  
17         99.4                1018.1    8.255408e+05                115.9  
58          NaN                   NaN    8.013060e+05                162.5  
110       100.6                 502.3    7.000366e+05                110.3  
121        75.7                 149.9    6.929690e+05                237.0  
18          NaN                   NaN    5.593588e+05                147.3  
132        96.5                 128.8    5.309382e+05               1466.9  
200        76.3                 179.1    5.126718e+05                  NaN  
137         NaN                   NaN    5.000079e+05                111.3  
27         59.4                 122.9    4.862625e+05                188.1  
136        65.3                 253.1    4.568483e+05                131.3  
292         NaN                   NaN    3.843544e+05                  NaN  
4          44.8                1769.5    3.562800e+05                105.9  
117        98.0                 205.6    3.362103e+05                  NaN  
122         NaN                   NaN    3.195293e+05                309.6  
5          35.4                 686.2    3.188856e+05                145.4  
267         NaN                   NaN    3.177201e+05                  NaN  
62          NaN                   NaN    3.066096e+05                137.9

In [ ]:
def seeValuesPotentialCompetitor (variable, country, number):
    if country == 'LVB':
        agg = df2.groupby('Item').mean()
        a = agg.sort_values(variable, ascending=False).head(number)
    else:
        a = df2[df2['Area'] == country].sort_values(variable, ascending=False).head(number)
    a.reset_index(inplace=True)
    a.index += 1
    a.rename(columns={'Item':'TopCrops'}, inplace=True)
    a.drop('Year', axis=1, inplace=True)
    return a

In [ ]:
def identifyPotentialCompetitor (variable, country, number):
    if country == 'LVB':
        agg = df2.groupby('Item').mean()
        agg.reset_index(inplace=True)
        topCrops = list(agg[['Item',variable]].sort_values(variable, ascending=False)['Item'].unique()[0:number])
    else:
        topCrops = list(df2[df2.Area == country][['Item', variable]].sort_values(variable, ascending=False)['Item'].unique()[0:number])
    return pd.DataFrame(topCrops, columns=['Top Crops'], index=range(1,number+1))
        
    

In [ ]:
a = identifyPotentialCompetitor('Food', 'RWA', 10)
a['Top Crops']

The potential competitor crops for rice could be:

You should first select among these variables: 'Food', 'Production', 'FoodPerCapita', 'FoodSelfSufficiency'
these areas: 'LVB', 'RWA', 'UGA', 'KEN', and 'TZA?
And then decide how many crops yuou want to get. 

In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.express as px

def graphCompetitors (variable, number):
    datos = df.copy()
    datos.Element.replace('Food supply quantity (kg/capita/yr)', 'FoodPerCapita', inplace=True)
    datos.Element.replace('Producer Price Index (2014-2016 = 100)', 'PriceIndex', inplace=True)
    datos.Element.replace('Gross Production Value (constant 2014-2016 million US$)','GrossProductionValue', inplace=True)
    datos = datos[(datos.Element == 'Food') & 
              ((datos.Category1 == 'Crop')|(datos.Category1 == 'CropProducts'))]
    datos = datos[['Domain', 'Area', 'Year', 'Element', 'Value', 'name', 'Unit']]
    datos = datos.rename(columns={'name' : 'Top Crops'})

    LVB = datos.groupby(['Domain', 'Year', 'Element', 'Top Crops', 'Unit']).mean()
    LVB['Area'] = 'LVB'
    LVB.reset_index(inplace=True)

    datos = datos.append(LVB[['Domain', 'Area', 'Year', 'Element', 'Value', 'Top Crops', 'Unit']])
        
    topCrops = identifyPotentialCompetitor(variable, 'LVB', number)
    datos = datos.merge(topCrops, how='inner', on='Top Crops')
    
    fig = px.line(datos,
                  x='Year', 
                  y = 'Value',
                  color='Top Crops', 
                  facet_col='Area',
                  facet_col_wrap=2,
                  title=variable, 
                  labels = {'Value': datos.Unit[1]})

    fig.show()

In [ ]:
graphCompetitors('Food', 20)

In [ ]:
seeValuesPotentialCompetitor('Production', 'LVB', 20)

In [ ]:
identifyPotentialCompetitor('FoodPerCapita', 'LVB', 15)

# Expansion of rice production in East Africa, and its implications on regional food security. 

- East African production is not sufficiency is 


## Changing dietary patterns in East Africa

Diets in East Africa are as heterogeneous as its agroecological zones. Sorghum and millet are widely cultivated and consumed in the semi-arid and arid regions, while platains and potatoes are cultivated in the humid highlands (see main food crops grown in Uganda in Figure 1). Maize, cassava, sweet potatoes, groundnuts, dark-green leaves, and rice are also important food crops in the region. Traditionally, cereal and grains were consumed in porridge ('uji' or 'ugali')  or in stews, and legumes, vegetables, and dark-green leaves were served as suplementary items. Cassava was seen as a reserve food, while meat was consumed during ceremonies or celebrations (Vorster, 1971). 

<br>
<div align="center">  <strong> Figure 1.   Main food crops in Uganda </strong> </div>

<div align="center"> <img src="images\StapleCropsUganda.png"
     alt="Markdown Monster icon"
     style="float: 5px; width:400px;"
     legend = "yep"/> </div>
     


<div align="center">  <em> Source: Uganda's Annual Agriculture Survey, 2018. </em> </div>
    
<br>    
However, dietary patterns have rapidly changed in the last 50 years. Traditional and coarse grains have been replaced by rice and wheat (Kennedy \& Reardon, 1994), consumption of processed foods and sugar have increased (Haggblade et al., 2016), and the amount of fruits and vegetables consumed has been insufficient to cover the micronutrient needs of the population (GNR, 2020). The following 4 figures show a decrease in the consumption of 'traditional' staples -sorghum, cassava, maize, and platains- and an increase in the consumption of rice in Kenya, Rwanda, Tanzania and Uganda. 


In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.express as px

fig = px.line(df[(df.Domain == 'FoodBalances') & 
                        (df.Element == 'Food') & 
                        ((df.Category1 == 'Crop')|(df.Category1 == 'CropProducts'))],
                 x='Year', 
                 y ='Value',
                 color='Item', 
                 facet_col='Area',
                 facet_col_wrap=2,
                 category_orders={'Area':['KEN', 'RWA', 'TZA', 'UGA']}, 
                 title='Food Production', 
                 labels = {'Value': 'Production [1000 tons]', 'Area': 'Country', 'Item': 'Crop'})

fig.show()
                         


In [ ]:
TZA.sort_values('Food', ascending=False).head(50)

### Uganda
Rice in Uganda is sold through three main channels: public markets, small grocery stores, and supermarkets. These channels are targeted to very differetn type of consumers. Rice is sold in packages, mainly in small stores and supermarkets, and loose in public markets. Despite the different varieties of rice that are produced in the country, in the market, rice is branded based on its origin, domestic and foreign productiction, not its variety. 

    
Rapid urbanization and rural-to-urban migration, mordernization of food retailing, and increases in income have been correlated with these dietary changes (Haggblade et al., 2016). A comparison of the 2005 and 2015 expenditure share in the food basket in Kenya showed a larger increase in consumption of rice in urban areas than in rural areas in Kenya (KIHBS, 2015). A similar trend is observed when comparing income groups; rice is consumed more by high-income clases since traditional crops, such as sorghum, are considered “inferior” foods (Minot, 2010). 

These dietary changes have negative impacts on public health. In African countries that have experienced the nutritional transition earlier than in East Africa, e.g., South Africa, it has been observed an increase in overweight and obesity, micronutrient deficiencies, and non-communicable diseases, such as diabetes and cadiovascular diseases (Steyn \& Mchiza, 2014). This can be partially explained by the substitution of whole-grains for refined grains and/or of traditional/diverse grains for standard-looking grains. But there are other processes influencing this, such as an increase in consumption of fast food and a decrease in consumption of vegetables and fruits. During the milling process, grains loss some vitamins that are contained in the bran and germ, such as vitamins B. For instance, brown rice and African rice (*Oryza glaberrima*) contain considerably more Thiamin, Riboflavin, Niacin, and Folate than white rice (see Table 1). The same can be observed when comparing whole-wheat with wheat flour. Similarly, yellow maize, which is associated with 'animal feed' and considered as 'poor-people's food' (IPBO, 2017), is richer in Vitamin A than white maize.   


**Table.1   Nutritional composition of main staple crops consumed in East Africa.**
    
|                                |     Rice, white, polished, raw | Rice, brown, raw |     Rice, red native, hulled, raw |     Maize, white, whole   kernel,   dried, raw |     Maize, yellow, whole   kernel,   dried, raw |     Wheat, whole grains, raw |     Wheat flour, white |     Sorghum, whole grain, raw  |     Pearl millet, whole grain,   raw   (with bran)    |     Potato, raw       |     Cassava, tuber, raw                     |     Plantain, ripe, raw |     Sweet potato, pale yellow, raw |     Sweet potato, deep yellow, raw |
|--------------------------------|--------------------------------|------------------|-----------------------------------|------------------------------------------------|-------------------------------------------------|------------------------------|------------------------|--------------------------------|-------------------------------------------------------|-----------------------|---------------------------------------------|-------------------------|------------------------------------|------------------------------------|
|     Scientific name            |     *Oryza sativa*               | *Oryza sativa*     |     *Oryza glaberrima*              |     Zea mays                                   |     Zea mays                                    |     Triticum spp.            |     Triticum spp.      |     Sorghum bicolor            |     Pennisetum glaucum                                |     Solanum tuberosum |     Manihot esculenta/   Manihot utilissima |     Musa paradisica     |     Ipomoea batatas                |     *Ipomea batatas*                 |
|     Edible   conversion factor |     1                          | 1                |     1                             |     1                                          |     1                                           |     1                        |     1                  |     1                          |     1                                                 |     0.84              |     0.84                                    |     0.65                |     0.84                           |     0.84                           |
|     Energy   kcal              |     353                        | 352              |     359                           |     349                                        |     353                                         |     326                      |     351                |     344                        |     364                                               |     80                |     153                                     |     140                 |     115                            |     112                            |
|     Protein   (g)              |     6.1                        | 7.8              |     7.4                           |     9.2                                        |     9                                           |     12.4                     |     10.4               |     10.5                       |     8.8                                               |     1.9               |     1.2                                     |     1.2                 |     1.5                            |     1.5                            |
|     Fat (g)                    |     0.5                        | 2.2              |     2.2                           |     4.1                                        |     4.5                                         |     2.2                      |     1.5                |     3.3                        |     5.8                                               |     0.1               |     0.3                                     |     0.3                 |     0.2                            |     0.3                            |
|     Fibre (g)                  | 0.4                            | 0.4              | 0.4                               |     9.7                                        |     9.7                                         |     10.8                     |     3.2                |     9.9                        |     9                                                 |     1.8               |     1.8                                     |     2.3                 |     3                              |     3                              |
|     Ca    (mg)                 |     11                         | 22               |     38                            |     19                                         |     12                                          |     30                       |     19                 |     24                         |     14                                                |     11                |     43                                      |     7                   |     27                             |     26                             |
|     Fe   (mg)                  |     0.7                        | 1.9              |     2.8                           |     3.1                                        |     3.5                                         |     4.7                      |     2                  |     3.7                        |     7.6                                               |     0.9               |     0.7                                     |     0.9                 |     1.1                            |     1.1                            |
|     Vit A-RAE   (mcg)          |     0                          | 0                |     0                             |     0                                          |     50                                          |     0                        |     0                  |     1                          |     0                                                 |     1                 |     1                                       |     43                  |     3                              |     397                            |
|     Thiamin (mg)               |     0.07                       | 0.38             |     0.34                          |     0.35                                       |     0.33                                        |     0.47                     |     0.28               |     0.36                       |     0.32                                              |     0.08              |     0.04                                    |     0.07                |     0.09                           |     0.09                           |
|     Riboflavin  (mg)           |     0.03                       | 0.07             |     0.1                           |     0.1                                        |     0.15                                        |     0.1                      |     0.1                |     0.16                       |     0.27                                              |     0.12              |     0.05                                    |     0.05                |     0.04                           |     0.05                           |
|     Niacin   (mg)              |     0.4                        | 5.0              |     6.5                           |     2.1                                        |     2.2                                         |     5.6                      |     1.2                |     3.3                        |     2.4                                               |     1.2               |     0.7                                     |     0.7                 |     0.6                            |     0.7                            |
|     Folate   (mcg)             |     20                         | 40               |                                   |     26                                         |     26                                          |     45                       |     24                 |     29                         |     30                                                |     18                |     24                                      |     22                  |     52                             |     52                             |


*Source:* FAO/INFOODS, (2012). Food Composition Table for Western Africa.
 
 
### Growing rice demand

Rice has grown in importance in East African countries, not only as food, but also as an economic opportunity. ... Something about rice consumption.... Rice serves both as a food staple for 70% of farming households and as a cash crop for 80% of these households ()... The share of total calories supplied by rice has been growing steadly since 1961 (Figure #). In the 60's, rice barely contributed to the kcal consumed in Kenya, Uganda, and Rwanda; yet, in 2013, it contributed with 2-5% of the calories consumed per capita in these countries. In Tanzania, where rice can be considered a staple, the share of calories coming from rice has increased from 3.7% in 1961 to 9.2% in 2013. 
    
**Figure 3. Percentage of total calories coming from main staple crops**. 

*Source: FAOSTAT*

Due to its popularity and demand, especially concerning aromatic rice, EAC countries import rice. Although Tanzania is largely self-sufficient and imports merely 140,000 MT rice per annum, other EAC import rice to varying degrees: Kenya imports 0.3 million MT and Rwanda imports 20 MT  (KilimoTrust, 2014a, p. 8). Rice is mainly imported from Asian countries (see Fig. 3), which is prized lower than rice produced in the region. The main rice exporters are Pakistan, Thailand, Tanzania, India and China, who combined exported a total of 1.1 million tons of rice in 2018 -an increase of 30% compared to year 2015 (UN Comtrade, 2020). Yet, trade within the EAC region plays a significant role in covering national demands; for example, Tanzania exported 82,300 t to Uganda in 2018 and 5,100 t to Burundi. Uganda also exported half of its production to other countries, such as Congo and South Sudan (UN Comtrade, 2020).   

...
    
#### Rice self-sufficiency 

Rice self-sufficiency has been rarely achieved in the last years in East African countries, with the exception of Tanzania. Although we have observed an increase in the harvested area and yields of paddy rice, domestic production has not been sufficient to cover the food demand for rice in Uganda, Rwanda, and Kenya as shown in Figure 4. In Rwanda, for instance, the production-consumption ratio (P/C) shows that self-sufficiency was achieved only once in the period between 1961-2013.     

**Figure 4. Rice harvested area, yields, imports and production/consumption ratio.** 